### PR TITLE
[bull] QueueOptions.redis cannot be a string

### PR DESCRIPTION
**Pull request by m-ronchi**

### CODE REVIEW

File: types/bull/index.d.ts [hunk 1]
```diff
@@ -55,7 +55,7 @@ declare namespace Bull {
     /**
      * Options passed directly to the `ioredis` constructor
      */
-    redis?: Redis.RedisOptions | string | undefined;
```
Comment (aleccool213) [inline types/bull/index.d.ts:58]: If these are truly options directly passed to ioredis (what the docstring says), you can pass a string (URI).

https://github.com/luin/ioredis#connect-to-redis

If this is incorrect, please update the doc string with that it does.

**Review by aleccool213 — CHANGES_REQUESTED** [review]

File: types/bull/index.d.ts (hunk 1 above)
Comment (m-ronchi) [reply types/bull/index.d.ts:58]: it used to be, but now bull clones the object before passing it to redis and overwrites `connectionName`

docs on bull about this are vague, there is only a comment in an example saying to use the object form ([here](https://github.com/OptimalBits/bull#basic-usage))

Comment (bondz) [reply types/bull/index.d.ts:58]: I don't think removing string here is accurate, Bull fully supports passing a string as optio. See [here](https://github.com/OptimalBits/bull/blob/c2194fead19e8d7a1159672e2f21de2a29d39990/lib/queue.js#L92-L100)

Comment (bondz) [reply types/bull/index.d.ts:58]: Oh. I think I was mistaken and you're right @m-ronchi. That was a different part of the code. When using an object, the redis key should not be a string. See [here](https://github.com/OptimalBits/bull/blob/c2194fead19e8d7a1159672e2f21de2a29d39990/lib/queue.js#L123-L130)

**Review by aleccool213 — APPROVED** [review]

**Comment by m-ronchi** [comment]
Ready to merge

**Comment by sorensenNiels** [comment]
This is a breaking change and if you look at ioredis documentation it states --->

You can also specify connection options as a redis:// URL or rediss:// URL when using TLS encryption:

// Connect to 127.0.0.1:6380, db 4, using password "authpassword":
new Redis("redis://:authpassword@127.0.0.1:6380/4");

// Username can also be passed via URI.
// It's worth to noticing that for compatibility reasons `allowUsernameInURI`
// need to be provided, otherwise the username part will be ignored.
new Redis(
  "redis://username:authpassword@127.0.0.1:6380/4?allowUsernameInURI=true"
);

**Comment by sorensenNiels** [comment]
Not sure how you can end up doing this again looking at the documentation of IORedit and the description of the constructor is clearly states you can define the connection using a string - please inlighten me!

**Comment by m-ronchi** [comment]
@sorensenNiels ioredis allows it, but bull does not since [v3.25](https://github.com/OptimalBits/bull/issues/2118), since it now mangles it assuming it is an object before passing it to redis.

if bull fixes it, you can revert this, but it seems [it won't](https://github.com/OptimalBits/bull/issues/2118#issuecomment-888211236)

tldr `new Queue('q', { redis: 'url' });` breaks at runtime
